### PR TITLE
Sepa and Giro pay

### LINKED ISSUE #313: Payment Details Page - Implement SEPA with Stripe
Allow Payment with SEPA direct debit for logged in users only.

SEPA credit transfer feature will be implemented in future to enable donations from guest users.

### PR DESCRIPTION
**Pull request by harshvitra**
Fixes #313 

- [x] Translations

- [x] Design of the GiroPay button

- [x] Condition for Sepa from payment Setup

- [x] Condition for GiroPay from payment Setup

- [x] Thank you page content

### **Testing hints -** 

-  Test Giropay. (it should only work when country is DE)

-  Test sepa 

-  Test createDonation API for all cases. 

Context - Now as soon as the user comes on the final page of donation we hit the createDonation API. Now when the user goes back and comes to the same page without any changes, we shouldn't hit the API, if the user makes some changes we should again call the createDonation API.

-  Test legacy Donations

### CODE REVIEW

**Comment by norbertschuler** [comment]
- [ ] I was able to test a Giropay donation now, but I am unsure what happens after the payment is done - does the user ever get the possibility download and share his donation as with the other payment methods? If we implemented the [extended functionality for user contributions](https://github.com/Plant-for-the-Planet-org/planet-webapp/issues/802), the user will probably be fine.

Bildschirmfoto 2021-02-09 um 15 50 28

**Comment by norbertschuler** [comment]
- [ ] If I click the button "Fail test payment" on the Stripe Giropay test page, I get shown the same Thank-You screen as when authorizing the payment:

Bildschirmfoto 2021-02-09 um 16 03 18

**Comment by harshvitra** [comment]
> * [ ]  What is the purpose of giving the donation id to the user?

@sagararyal 

Also do we share donation ID for all payment methods or just for Giropay ?

**Comment by norbertschuler** [comment]
> * Test Giropay. (it should only work when country is DE)

Already tested above (see comments):
- Missing a special case if payment has failed.
- Missing sharing options - maybe omitted intentionally?.

> * Test sepa

- [ ] Currently allowing IBANs from all countries independent from user country - I think that's ok to allow the user be be more flexible?

> * Test createDonation API for all cases.
> 
> Context - Now as soon as the user comes on the final page of donation we hit the createDonation API. Now when the user goes back and comes to the same page without any changes, we shouldn't hit the API, if the user makes some changes we should again call the createDonation API.

- [ ] Avoidance of duplicate API calls for creating a donation is not yet working. I always see a new API call to create a donation going back to the address form and forward to the payment form again without any changes. 

> * Test legacy Donations

- [ ] Legacy donations do not support SEPA or Giropay, right? I only just got the option to pay with my credit card for a project which supports SEPA and Giropay in in the webapps:

Bildschirmfoto 2021-02-10 um 09 07 30

- [ ] The backend API call to do the legacy donation then is failing with: `"Call to a member function getUserProfile() on null\nfile: /app/src/Common/Payment/Service/PaymentService.php, line: 53"`

```
Request URL: https://app-staging.plant-for-the-planet.org/app/donations/don_K4XPpnB9LHaSn1atW85GWHR6
Request Method: PUT
Status Code: 500 
Payload: {"paymentProviderRequest":{"account":"acct_1DYCMDD2OpW2f42N","gateway":"stripe_pi","source":{"id":"pm_1IJDkLD2OpW2f42N6lkBHYIt","object":"payment_method"}}}
```

**Comment by harshvitra** [comment]
> * Missing sharing options - maybe omitted intentionally?.
> 
> >

Sharing options are now available

**Comment by norbertschuler** [comment]
> > * Missing sharing options - maybe omitted intentionally?.
> > 
> > >
> 
> Sharing options are now available

O.k. - as you removed the donation id the question about it's purpose it obsolete now ;-)

Bildschirmfoto 2021-02-10 um 11 24 06

**Comment by harshvitra** [comment]
Pending and Failed components 

Screenshot 2021-02-11 at 3 59 56 PM

Screenshot 2021-02-11 at 3 59 23 PM

**Comment by norbertschuler** [comment]
> Pending and Failed components

- [ ] Great, but currently credit cards payment are now getting a `paymentStatus: "pending" state when I am testing. Any idea?

**Comment by norbertschuler** [comment]
- [ ] Giropay works for me (choosing project "GEO - Cuellaje Ecuador von GEO schützt den Regenwald e.V."), but I am unsure what will happen if users go back in the browser and hit the payment API several times (e.g. go back to the Giropay test page and click "Authorize" or "Fail payment" again). @sagararyal Where can I see the Giropay test payments?

- [ ] For the payment method "sofort" (same project) I get the following error on the staging backend:

```
Request URL: https://app-staging.plant-for-the-planet.org/app/donations/don_3UGc5zFme5m0wi1iYU4CJ9tv
Request Method: PUT
Status Code: 202 
{"id":"don_3UGc5zFme5m0wi1iYU4CJ9tv","status":"failed","errorCode":"parameter_unknown","message":"Received unknown parameter: sofort"}

```
Bildschirmfoto 2021-02-12 um 09 47 44

**Comment by norbertschuler** [comment]
- [ ] I tested a SEPA payment which resulted into the following screen - I wondered if we ever want to offer the user paying with SEPA also a Thank-You screen? The wording says "feel free to close this page", but if the user waits, no additional API request would be made to check if the payment got finalized and show any thank you page, right?

Bildschirmfoto 2021-02-12 um 09 58 20

**Comment by norbertschuler** [comment]
- [ ] For the payment method "sofort" (same project) I get the following error on the staging backend:

```
Request URL: https://app-staging.plant-for-the-planet.org/app/donations/don_3UGc5zFme5m0wi1iYU4CJ9tv
Request Method: PUT
Status Code: 202 
{"id":"don_3UGc5zFme5m0wi1iYU4CJ9tv","status":"failed","errorCode":"parameter_unknown","message":"Received unknown parameter: sofort"}

```
Bildschirmfoto 2021-02-12 um 09 47 44

File: src/features/donations/components/paymentMethods/CardPayments.tsx [hunk 1]
```diff
@@ -101,7 +102,20 @@ function CardPayments({
   }, [CardNumberElement, CardExpiryElement, CardCvcElement]);
 
   const createPaymentMethodCC = (cardElement: any) => {
-    return stripe.createPaymentMethod('card', cardElement)
+    return stripe.createPaymentMethod({
+      type: 'card',
+      card: cardElement,
+      billing_details: {
+        name: `${donorDetails.firstname} ${donorDetails.lastname}`,
```
Comment (norbertschuler) [inline src/features/donations/components/paymentMethods/CardPayments.tsx:109]: Legacy payments fail here with the following error as probably the users name is not available (I only see the API call https://app-staging.plant-for-the-planet.org/public/v1.3/de/paymentInfo/JFCOX8DBEI - nothing loading a user name):
```
CardPayments.tsx:109 Uncaught (in promise) TypeError: Cannot read property 'firstname' of undefined
```
Tested with https://planet-webapp-kjunubc87.vercel.app/donation-payment/JFCOX8DBEI

**Review by norbertschuler — CHANGES_REQUESTED** [review]

File: src/features/donations/components/paymentMethods/CardPayments.tsx (hunk 1 above)
Comment (norbertschuler) [reply src/features/donations/components/paymentMethods/CardPayments.tsx:109]: I did a donation in the Treecounter app without being logged in just entering a random address.

Comment (norbertschuler) [reply src/features/donations/components/paymentMethods/CardPayments.tsx:109]: donorDetails is not probably not a mandatory property for CardPayments - could such an error be prevented by type checking?

Comment (harshvitra) [reply src/features/donations/components/paymentMethods/CardPayments.tsx:109]: This was added to pass billing details to stripe.

@sagararyal we will either have to remove the billing details from CC method or add the user's details in this API - `https://app-staging.plant-for-the-planet.org/public/v1.3/en/paymentInfo/JFCOX8DBEI`

Comment (harshvitra) [reply src/features/donations/components/paymentMethods/CardPayments.tsx:109]: I have pushed a condition, if donor details are not present we don't pass the billing details

Comment (norbertschuler) [reply src/features/donations/components/paymentMethods/CardPayments.tsx]: Now the API call can be triggered.

**Comment by norbertschuler** [comment]
- [ ] After sending a legacy donation, the empty payment form is shown again instead of the thank you page:

Bildschirmfoto 2021-02-12 um 13 22 47

**Comment by sagararyal** [comment]
> [ ] Giropay works for me (choosing project "GEO - Cuellaje Ecuador von GEO schützt den Regenwald e.V."), but I am unsure what will happen if users go back in the browser and hit the payment API several times (e.g. go back to the Giropay test page and click "Authorize" or "Fail payment" again). @sagararyal Where can I see the Giropay test payments?

@norbertschuler This can be checked in Backend or Stripe. Giropay can't be authorized twice for the same payment Intent. 

Once a payment Intent is successful or failed, a new payment intent has to be created.

**Pull request merged**